LLVM9 debuginfo change

RichBarton-Arm opened the pull request:
LLVM9 Changes
LLVM9 moves some bitfields from the DIFlags to a subprogram-specific field called DISPFlags. Teach flang2 to emit these, whilst staying backwards compatible.

gklimowicz commented:
@jiel-nv Can you look these over for us?

jiel-nv commented on tools/flang2/flang2exe/lldebug.cpp:
```diff
@@ -181,11 +189,20 @@ void
 {
 #ifdef FLANG_LLVM_EXTENSIONS
-  if (ll_feature_debug_info_ver70(feature)) {
+  DIFLAG_ARTIFICIAL = 1 << 6;
```
It seems to me the DIFLAG_ARTIFICIAL = 1 << 6 should be pulled out of the FLANG_LLVM_EXTENSIONS guard.

jiel-nv commented on tools/flang2/flang2exe/lldebug.cpp:
```diff
@@ -447,18 +487,23 @@ lldbg_create_subprogram_mdnode(
-  llmd_add_i1(mdb, is_optimized);
+  if (ll_feature_debug_info_ver90(&db->module->ir))
+    llmd_add_i32(mdb, spFlags);
+  if (!ll_feature_debug_info_ver90(&db->module->ir))
```
Could be an "else" branch?

schweitzpgi replied:
An `else` would make this a tad clearer to read. (The `!` is less clear.)

RichBarton-Arm replied:
Ack

schweitzpgi commented on tools/flang2/flang2exe/lldebug.cpp:
```diff
@@ -75,12 +75,20 @@ inline SPTR GetParamSptr(int dpdsc, int i) {
+static int DISPFLAG_PURE;
+static int DISPFLAG_ELEMENTAL;
+static int DISPFLAG_RECUSIVE;
+static int DISPFLAG_ISMAINPGM;
```
These flags should all be `const` since they do not change in prior versions of LLVM IR. (They didn't exist or were not used if they did, so there is no backward compatibility issue.)

schweitzpgi commented on tools/flang2/flang2exe/lldebug.cpp:
```diff
@@ -195,6 +212,29 @@ InitializeDIFlags(const LL_IRFeatures *feature)
+    DISPFLAG_ISMAINPGM = 0;
+  }
+}
+
```
This function isn't needed if we make the DISP flags constants.

schweitzpgi commented on tools/flang2/flang2exe/lldebug.cpp:
```diff
@@ -2008,23 +2064,27 @@ lldbg_emit_outlined_subprogram(LL_DebugInfo *db, int sptr, int findex,
   is_def = DEFDG(sptr);
   is_def |= (STYPEG(sptr) == ST_ENTRY);
+  if (ll_feature_has_diextensions(&db->module->ir))
+    flags = set_disubprogram_flags(sptr);
```
I think it would be clearer if `set_disubprogram_flags` was customized to omit the DIFLAGs that are no longer in that flags word rather than introducing more pseudo-constants and relying on them being reset to 0.

RichBarton-Arm commented:
Thanks @jiel-nv and @schweitzpgi for the reviews. 

I have moved all the flag values back to being const variables. The ones that changed value between LLVM versions are now different variables. The functions that initialse them are gone and instead the function that sets the flag value uses the right values for the LLVM version being generated for.

I also fixed the small comment on else vs !

I have uploaded a third commit which should address them all, but I am not sure if you would prefer me to squash the commits together before the final pull. Let me know if you want me to do this.

sscalpone commented:
Yes, please squash.

RichBarton-Arm commented:
Seems like this got merged in in other commits, so closing.
